pkg/store/bucket.go

```diff
@@ -1101,64 +1199,63 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storepb.Store_Serie
 		}
 
 		for _, b := range blocks {
-			b := b
+			blk := b
 			gctx := gctx
 
 			if s.enableSeriesResponseHints {
 				// Keep track of queried blocks.
-				resHints.AddQueriedBlock(b.meta.ULID)
+				resHints.AddQueriedBlock(blk.meta.ULID)
 			}
 
-			var chunkr *bucketChunkReader
-			// We must keep the readers open until all their data has been sent.
-			indexr := b.indexReader()
-			if !req.SkipChunks {
-				chunkr = b.chunkReader()
-				defer runutil.CloseWithLogOnErr(s.logger, chunkr, "series block")
-			}
-
-			// Defer all closes to the end of Series method.
-			defer runutil.CloseWithLogOnErr(s.logger, indexr, "series block")
+			shardMatcher := req.ShardInfo.Matcher(&s.buffers)
+			blockClient := newBlockSeriesClient(
+				srv.Context(),
+				s.logger,
+				blk,
+				req,
+				chunksLimiter,
+				bytesLimiter,
+				shardMatcher,
+				s.enableChunkHashCalculation,
+				s.seriesBatchSize,
+				s.metrics.chunkFetchDuration,
+			)
+			defer blockClient.Close()
 
 			g.Go(func() error {
-				span, newCtx := tracing.StartSpan(gctx, "bucket_store_block_series", tracing.Tags{
-					"block.id":         b.meta.ULID,
-					"block.mint":       b.meta.MinTime,
-					"block.maxt":       b.meta.MaxTime,
-					"block.resolution": b.meta.Thanos.Downsample.Resolution,
+				span, _ := tracing.StartSpan(gctx, "bucket_store_block_series", tracing.Tags{
+					"block.id":         blk.meta.ULID,
+					"block.mint":       blk.meta.MinTime,
+					"block.maxt":       blk.meta.MaxTime,
+					"block.resolution": blk.meta.Thanos.Downsample.Resolution,
 				})
-				defer span.Finish()
-
-				shardMatcher := req.ShardInfo.Matcher(&s.buffers)
-				defer shardMatcher.Close()
```

Comment:
Did we close the matcher properly in this pr? I didn't see where we close it after the refactoring

Reply:
The response set should close it here: https://github.com/thanos-io/thanos/blob/main/pkg/store/proxy_heap.go#L590

Are you seeing any issues with this change?